[Clang] Fix Microsoft ABI inheritance model when member pointer is used in a base specifier

MitalAshok opened the pull request:
Fix `CXXRecordDecl::isParsingBaseSpecifiers` so that it is true while parsing base specifiers instead of directly after they have been parsed.

`-fcomplete-member-pointers` now issues a diagnostic when a member pointer is used in a base specifier.

MitalAshok committed:
[Clang] Fix Microsoft ABI inheritance model when member pointer is used in a base specifier

Fix CXXRecordDecl::isParsingBaseSpecifiers so that it is true while parsing base specifiers instead of directly after they have been parsed.

-fcomplete-member-pointers now issues a diagnostic when a member pointer is used in a base specifier.

-fcomplete-member-pointers has also been relaxed to not issue a diagnostic for incomplete classes with an explicit __{single|multiple|virtual}_inheritance attribute, whose completeness would not affect the representation of pointer-to-member objects.

MitalAshok commented:
Example of the incompatibility: https://godbolt.org/z/Mn1T57WGb

```c++
struct unspecified_inheritance;
template<int I>
struct X {
    static_assert(I == sizeof(int unspecified_inheritance::*), "");
};

struct Y : X<sizeof(int Y::*)> {};
```

Currently, the `sizeof(int Y::*)` locks down the inheritance model of `Y` as single, when MSVC marks it as unspecified.

It was the intent that it should have been unspecified:

https://github.com/llvm/llvm-project/blob/89a080cb79972abae240c226090af9a3094e2269/clang/lib/AST/MicrosoftCXXABI.cpp#L223-L224

But `isParsingBaseSpecifiers()` was set after parsing instead of before.

MitalAshok committed:
Merge branch 'main' into ms-member-pointer-in-base

MitalAshok committed:
Add more places where -fcomplete-member-pointers should match the Microsoft ABI

MaxEW707 commented on clang/include/clang/AST/DeclCXX.h:
```diff
@@ -598,7 +599,9 @@ class CXXRecordDecl : public RecordDecl {
   }
 
-  void setIsParsingBaseSpecifiers() { data().IsParsingBaseSpecifiers = true; }
+  void setIsParsingBaseSpecifiers(bool to = true) {
```
```suggestion
  void setIsParsingBaseSpecifiers(bool To = true) {
```

I am surprised clang-format didn't complain about starting with a lowercase for variables.
I can't think of a better name. I respond here if I do :).

MitalAshok committed:
Merge branch 'main' into ms-member-pointer-in-base

MitalAshok committed:
Revert "Add more places where -fcomplete-member-pointers should match the Microsoft ABI"

This reverts commit f3436707c24abac102a4f68e01b5dd1fe6dc334c.

I'm planning on moving this to a separate pr which substantially changes -fcomplete-member-pointers

MitalAshok committed:
Partially revert changes to -fcomplete-member-pointers

I'm planning on moving this to a separate pr which substantially changes -fcomplete-member-pointers

MitalAshok committed:
Manipulate the CXXRecordDecl in Sema instead of Parse

Endilll reviewed (commented):
`Sema.h` changes look good.

MaxEW707 reviewed (approved):
LGTM.

@MitalAshok I noticed in your commit messages you mentioned reworking -fcomplete-member-pointers. Will that include fixing clang incorrectly warning when an inheritance model is explicitly specified? Godbolt for reference: https://godbolt.org/z/rYaWz6sra.

MitalAshok commented:
> Will that include fixing clang incorrectly warning when an inheritance model is explicitly specified?

Yes, I have a WIP of it here #98010 where that doesn't get a warning (https://github.com/llvm/llvm-project/pull/98010/files#diff-bec1ae40e5793adbd61e7a9dc41a4715a06e27888a384c0752bb6b4b09a1075fR20-R21)

MaxEW707 commented:
@Endilll Wondering if this PR gets the green check from you.

I would like to get this merged before the clang-19 branch is created and just want to get the green light from a more senior member of the community before merging. Thanks :).

cor3ntin commented:
@AaronBallman

AaronBallman commented on clang/lib/Sema/SemaType.cpp:
```diff
@@ -9015,11 +9015,20 @@ bool Sema::RequireCompleteTypeImpl(SourceLocation Loc, QualType T,
-        return true;
+      if (getLangOpts().CompleteMemberPointers) {
+        const CXXRecordDecl *RD = MPTy->getClass()->getAsCXXRecordDecl();
+        if (RD->isBeingDefined()) {
+          if (RD->isParsingBaseSpecifiers()) {
```
Combine these into a single statement?

MitalAshok replied:
I really only want to call `RequireCompleteType`in the else branch  when `!RD->isBeingDefined()` (when called while `RD->isParsingBaseSpecifiers()`, there's an extra note 'definition of "[class name]" is incomplete until the closing "}"', which is misleading since for CompleteMemberPointers purposes it's complete after the bases are parsed)

rnk replied:
I'm struggling to understand why this error diagnostic doesn't fire in your MS ABI test case example that forms a member pointer type in the base specifier list.

AaronBallman reviewed (commented):
This generally LGTM but I'd love to hear from @rnk or @zmodem as they are quite knowledgeable about the MS ABI in general.

rnk commented on clang/test/SemaCXX/complete-member-pointers.cpp:
```diff
@@ -13,3 +15,18 @@ template <typename T>
+S5 // #S5
+:
+Base<sizeof(int S5::*)>
+// expected-error@-1 {{member pointer has incomplete base type 'S5'}}
```
Should this be an error in Microsoft mode? Shouldn't we silently use the unspecified model in that mode?

MitalAshok replied:
It's an error for `-fcomplete-member-pointers` which is supposed to tell you when you use a member pointer in a way that the MS ABI would have it be unspecified, regardless of if you are compiling with MS member pointers or Itanium member pointers

MitalAshok replied:
Yes, that's just how `-fcomplete-member-pointers` works currently. It's more like a `-Werror=...` diagnostic

rnk commented on clang/test/SemaCXX/member-pointer-ms.cpp:
```diff
@@ -215,6 +215,14 @@ struct NewUnspecified;
 UnspecTemplate<void (NewUnspecified::*)()> tmpl_unspec;
 
+// Member pointers used in base specifiers force an unspecified inheritance model
+struct MemPtrInBase : UnspecTemplate<void (MemPtrInBase::*)()> {};
```
I guess this memptr type is not required to be complete during the parsing of the base specifiers here.

MitalAshok replied:
It's not for simply naming the type for the template argument but `UnspecTemplate`'s definition does complete it by taking its `sizeof`, forcing the inheritance model to be assigned

MitalAshok replied:
I see what you mean.  
If there were two bases `struct MemPtrInBase : UnspecTemplate<void (MemPtrInBase::*)()>, SomeOtherClass {`, `UnspecTemplate<void (MemPtrInBase::*)()>` would be instantiated and completed before the token for `SomeOtherClass` is seen. So this is the case for Clang, but I don't see why a different compiler couldn't parse all the bases at once then instantiated them.

However, they need to be instantiated at least before the body of the struct is parsed (because the bases affect name lookup), and this is essentially the same time as when you are "parsing the base specifiers".  
And of course it needs to match Microsoft behaviour and assign the unspecified inheritance model during this time.

MaxEW707 commented:
@rnk @zmodem Sorry for the ping. Its been a bit since this got reviewed and I recently hit this ABI case when moving some parts of a larger project to clang.

rnk reviewed (commented):
Sorry, looks like I wrote pending comments and didn't publish them
